### PR TITLE
Introduce Background Compatibility for Socket Mode

### PR DESCRIPTION
**Pull request by huseyinbabal**
## Description

Changes proposed in this pull request:

- Added `botToken` and `appToken` for custom slack app
- `token` can be used for `Add to Slack` feature

## Testing Locally
E2e tests in CI became flaky since most probably we use ci app on our locals and what we sent as a botkube command, it is responded from another computer other than existing ci e2e. To be on the safe side, I tested everything from scratch as shown below.

### Create slack app
Create a slack app in our CI workspace with following manifest
```
display_information:
  name: Botkubev3
  description: Botkubev3
  background_color: "#a653a6"
features:
  bot_user:
    display_name: Botkubev3
    always_online: false
oauth_config:
  scopes:
    bot:
      - app_mentions:read
      - chat:write
      - channels:read
      - files:write
settings:
  event_subscriptions:
    bot_events:
      - app_mention
  interactivity:
    is_enabled: true
  org_deploy_enabled: false
  socket_mode_enabled: true
  token_rotation_enabled: false
```

### Collect app and bot tokens as shown
```
export SLACK_BOT_TOKEN=
export SLACK_APP_TOKEN=
export SLACK_TESTER_APP_TOKEN=
```

### Run Botkube Deployment with Test Enabled
```
helm install botkube -n botkube --create-namespace \
-f ./helm/botkube/e2e-test-values.yaml \
--set image.repository=kubeshop/pr/botkube \
--set image.tag=732-PR \
--set e2eTest.image.repository=kubeshop/pr/botkube-test \
--set e2eTest.image.tag=732-PR \
--set e2eTest.slack.botName="botkubev3" \
--set e2eTest.slack.testerAppToken="${SLACK_TESTER_APP_TOKEN}" \
--set communications.default-group.slack.appToken="${SLACK_APP_TOKEN}" \
--set communications.default-group.slack.botToken="${SLACK_BOT_TOKEN}" \
./helm/botkube
```

### Run tests
```
helm test botkube --namespace botkube --timeout=10m --logs
```

### Here is my local output
```
helm test botkube --namespace botkube --timeout=10m --logs                                                                                    [5dd817b]
NAME: botkube
LAST DEPLOYED: Thu Sep 15 21:00:26 2022
NAMESPACE: botkube
STATUS: deployed
REVISION: 1
TEST SUITE:     botkube-e2e-test
Last Started:   Thu Sep 15 21:00:29 2022
Last Completed: Thu Sep 15 21:06:28 2022
Phase:          Succeeded

POD LOGS: botkube-e2e-test
=== RUN   TestRemoveSlackLinksIndicators
=== RUN   TestRemoveSlackLinksIndicators/no_links
=== RUN   TestRemoveSlackLinksIndicators/one_link
=== RUN   TestRemoveSlackLinksIndicators/multiple_links
--- PASS: TestRemoveSlackLinksIndicators (0.00s)
    --- PASS: TestRemoveSlackLinksIndicators/no_links (0.00s)
    --- PASS: TestRemoveSlackLinksIndicators/one_link (0.00s)
    --- PASS: TestRemoveSlackLinksIndicators/multiple_links (0.00s)
=== RUN   TestSlack
    slack_test.go:60: Loading configuration...
    slack_test.go:65: Creating Slack API client with provided token...
    slack_test.go:69: Creating K8s client...
W0915 18:00:30.975927       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
    slack_test.go:75: Setting up test Slack setup...
    slack_tester_test.go:94: Getting users...
    slack_tester_test.go:98: Finding user ID by name "botkubev3"...
    slack_tester_test.go:94: Getting users...
    slack_tester_test.go:98: Finding user ID by name "botkube_tester"...
    slack_test.go:79: Creating channel "test-493f2406-d569-4208-aa8a-6c6e07295adc"...
    slack_test.go:79: Channel "test-493f2406-d569-4208-aa8a-6c6e07295adc" (ID: "C043A6LQUHW") created
    slack_test.go:81: Creating channel "test-93721fcd-25cc-4cfe-9d97-873d6e4fb5ac"...
    slack_test.go:81: Channel "test-93721fcd-25cc-4cfe-9d97-873d6e4fb5ac" (ID: "C042LC9LZM1") created
    slack_test.go:89: Posting welcome message...
    slack_tester_test.go:110: Inviting bot with ID "U043A5M2FNU" to the channel with ID "C043A6LQUHW"
    slack_test.go:89: Posting welcome message...
    slack_tester_test.go:110: Inviting bot with ID "U043A5M2FNU" to the channel with ID "C042LC9LZM1"
    slack_test.go:93: Patching Deployment with test env variables...
    slack_test.go:98: Waiting for Deployment
    slack_test.go:102: Waiting for Bot message on channel...
    slack_test.go:108: Running actual test cases
=== RUN   TestSlack/Ping
=== RUN   TestSlack/Filters_list
=== RUN   TestSlack/Commands_list
=== RUN   TestSlack/Commands_list/With_default_cluster
=== RUN   TestSlack/Commands_list/With_custom_cluster_name
=== RUN   TestSlack/Commands_list/With_unknown_cluster_name
    slack_test.go:205: Ensuring bot didn't write anything new...
=== RUN   TestSlack/Executor
=== RUN   TestSlack/Executor/Get_Deployment
=== RUN   TestSlack/Executor/Get_Configmap
=== RUN   TestSlack/Executor/Get_forbidden_resource
=== RUN   TestSlack/Executor/Specify_unknown_command
=== RUN   TestSlack/Executor/Specify_invalid_command
=== RUN   TestSlack/Executor/Specify_forbidden_namespace
=== RUN   TestSlack/Executor/Based_on_other_bindings
=== RUN   TestSlack/Executor/Based_on_other_bindings/Wait_for_Deployment_(the_2st_binding)
=== RUN   TestSlack/Executor/Based_on_other_bindings/Exec_(the_3rd_binding_which_is_disabled)
=== RUN   TestSlack/Executor/Based_on_other_bindings/Get_all_Pods_(the_4th_binding)
=== RUN   TestSlack/Executor/Based_on_other_bindings/Get_all_Deployments_(the_4th_binding)
=== RUN   TestSlack/Multi-channel_notifications
    slack_test.go:335: Creating ConfigMap...
    slack_test.go:348: Expecting bot message in first channel...
    slack_test.go:360: Expecting no bot message in second channel...
    slack_test.go:366: Updating ConfigMap...
    slack_test.go:373: Expecting bot message in all channels...
    slack_test.go:385: Stopping notifier...
    slack_test.go:393: Getting notifier status from second channel...
    slack_test.go:400: Getting notifier status from first channel...
    slack_test.go:407: Updating ConfigMap once again...
    slack_test.go:414: Ensuring bot didn't write anything new on first channel...
    slack_test.go:420: Expecting bot message on second channel...
    slack_test.go:431: Starting notifier
    slack_test.go:438: Creating and deleting ignored ConfigMap
    slack_test.go:453: Ensuring bot didn't write anything new...
    slack_test.go:458: Deleting ConfigMap
    slack_test.go:463: Expecting bot message on first channel...
    slack_test.go:475: Ensuring bot didn't write anything new on second channel...
=== RUN   TestSlack/Recommendations
    slack_test.go:492: Creating Pod...
    slack_test.go:510: Expecting bot message...
    slack_test.go:569: Cleaning up created Pod...
=== CONT  TestSlack
    slack_test.go:96: Restoring envs for deployment...
    slack_test.go:82: Archiving channel "test-93721fcd-25cc-4cfe-9d97-873d6e4fb5ac"...
    slack_test.go:80: Archiving channel "test-493f2406-d569-4208-aa8a-6c6e07295adc"...
--- PASS: TestSlack (355.58s)
    --- PASS: TestSlack/Ping (1.52s)
    --- PASS: TestSlack/Filters_list (1.60s)
    --- PASS: TestSlack/Commands_list (64.55s)
        --- PASS: TestSlack/Commands_list/With_default_cluster (1.50s)
        --- PASS: TestSlack/Commands_list/With_custom_cluster_name (1.48s)
        --- PASS: TestSlack/Commands_list/With_unknown_cluster_name (61.57s)
    --- PASS: TestSlack/Executor (17.19s)
        --- PASS: TestSlack/Executor/Get_Deployment (1.51s)
        --- PASS: TestSlack/Executor/Get_Configmap (1.60s)
        --- PASS: TestSlack/Executor/Get_forbidden_resource (1.55s)
        --- PASS: TestSlack/Executor/Specify_unknown_command (2.51s)
        --- PASS: TestSlack/Executor/Specify_invalid_command (2.55s)
        --- PASS: TestSlack/Executor/Specify_forbidden_namespace (1.52s)
        --- PASS: TestSlack/Executor/Based_on_other_bindings (5.96s)
            --- PASS: TestSlack/Executor/Based_on_other_bindings/Wait_for_Deployment_(the_2st_binding) (1.50s)
            --- PASS: TestSlack/Executor/Based_on_other_bindings/Exec_(the_3rd_binding_which_is_disabled) (1.51s)
            --- PASS: TestSlack/Executor/Based_on_other_bindings/Get_all_Pods_(the_4th_binding) (1.47s)
            --- PASS: TestSlack/Executor/Based_on_other_bindings/Get_all_Deployments_(the_4th_binding) (1.48s)
    --- PASS: TestSlack/Multi-channel_notifications (261.63s)
    --- PASS: TestSlack/Recommendations (1.27s)
PASS
```

### CODE REVIEW

File: pkg/bot/slack.go [hunk 1]
```diff
@@ -67,8 +67,13 @@ type slackMessage struct {
 
 // NewSlack creates a new Slack instance.
 func NewSlack(log logrus.FieldLogger, cfg config.Slack, executorFactory ExecutorFactory, reporter FatalErrorAnalyticsReporter) (*Slack, error) {
-	client := slack.New(cfg.Token, slack.OptionAppLevelToken(SlackAppLevelToken))
-
```
Comment (mszostok) [inline pkg/bot/slack.go:71]: Please introduce some logging around that and also validation.

Here I create my proposition:  https://github.com/kubeshop/botkube/commit/5453531488df238a8a14f2a65b736005c0c166cd



I was not able to do the validation via tags, so I registered a dedicated function. IMO we should move to JSON Schema validation at some point.

Comment (huseyinbabal) [reply pkg/bot/slack.go:71]: sure

**Comment by huseyinbabal** [comment]
@mszostok thanks for the comments, I applied them and waiting for docker images, will run tests locally

**Comment by huseyinbabal** [comment]
```
2022/09/16 06:47:02 while loading app configuration: found critical validation errors: 1 error occurred:                                               │
│     * Key: 'Config.Communications[default-group].Slack.Channels' Channels is not equal to 1
```
Got this error, on it.

**Comment by huseyinbabal** [comment]
Ok fixed and tested locally
```
→ helm test botkube --namespace botkube --timeout=10m --logs                                                                                    [18dd633]
NAME: botkube
LAST DEPLOYED: Fri Sep 16 10:13:07 2022
NAMESPACE: botkube
STATUS: deployed
REVISION: 1
TEST SUITE:     botkube-e2e-test
Last Started:   Fri Sep 16 10:14:25 2022
Last Completed: Fri Sep 16 10:20:21 2022
Phase:          Succeeded

POD LOGS: botkube-e2e-test
=== RUN   TestRemoveSlackLinksIndicators
=== RUN   TestRemoveSlackLinksIndicators/no_links
=== RUN   TestRemoveSlackLinksIndicators/one_link
=== RUN   TestRemoveSlackLinksIndicators/multiple_links
--- PASS: TestRemoveSlackLinksIndicators (0.00s)
    --- PASS: TestRemoveSlackLinksIndicators/no_links (0.00s)
    --- PASS: TestRemoveSlackLinksIndicators/one_link (0.00s)
    --- PASS: TestRemoveSlackLinksIndicators/multiple_links (0.00s)
=== RUN   TestSlack
    slack_test.go:60: Loading configuration...
    slack_test.go:65: Creating Slack API client with provided token...
    slack_test.go:69: Creating K8s client...
W0916 07:14:27.149657       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
    slack_test.go:75: Setting up test Slack setup...
    slack_tester_test.go:94: Getting users...
    slack_tester_test.go:98: Finding user ID by name "botkubev3"...
    slack_tester_test.go:94: Getting users...
    slack_tester_test.go:98: Finding user ID by name "botkube_tester"...
    slack_test.go:79: Creating channel "test-8daaab80-f7b7-4cce-9dce-6e883967ee4e"...
    slack_test.go:79: Channel "test-8daaab80-f7b7-4cce-9dce-6e883967ee4e" (ID: "C0428J27P4P") created
    slack_test.go:81: Creating channel "test-11118aef-682b-41e8-b800-a7c6fdeba05b"...
    slack_test.go:81: Channel "test-11118aef-682b-41e8-b800-a7c6fdeba05b" (ID: "C0428J28P7H") created
    slack_test.go:89: Posting welcome message...
    slack_tester_test.go:110: Inviting bot with ID "U043A5M2FNU" to the channel with ID "C0428J27P4P"
    slack_test.go:89: Posting welcome message...
    slack_tester_test.go:110: Inviting bot with ID "U043A5M2FNU" to the channel with ID "C0428J28P7H"
    slack_test.go:93: Patching Deployment with test env variables...
    slack_test.go:98: Waiting for Deployment
    slack_test.go:102: Waiting for Bot message on channel...
    slack_test.go:108: Running actual test cases
=== RUN   TestSlack/Ping
=== RUN   TestSlack/Filters_list
=== RUN   TestSlack/Commands_list
=== RUN   TestSlack/Commands_list/With_default_cluster
=== RUN   TestSlack/Commands_list/With_custom_cluster_name
=== RUN   TestSlack/Commands_list/With_unknown_cluster_name
    slack_test.go:205: Ensuring bot didn't write anything new...
=== RUN   TestSlack/Executor
=== RUN   TestSlack/Executor/Get_Deployment
=== RUN   TestSlack/Executor/Get_Configmap
=== RUN   TestSlack/Executor/Get_forbidden_resource
=== RUN   TestSlack/Executor/Specify_unknown_command
=== RUN   TestSlack/Executor/Specify_invalid_command
=== RUN   TestSlack/Executor/Specify_forbidden_namespace
=== RUN   TestSlack/Executor/Based_on_other_bindings
=== RUN   TestSlack/Executor/Based_on_other_bindings/Wait_for_Deployment_(the_2st_binding)
=== RUN   TestSlack/Executor/Based_on_other_bindings/Exec_(the_3rd_binding_which_is_disabled)
=== RUN   TestSlack/Executor/Based_on_other_bindings/Get_all_Pods_(the_4th_binding)
=== RUN   TestSlack/Executor/Based_on_other_bindings/Get_all_Deployments_(the_4th_binding)
=== RUN   TestSlack/Multi-channel_notifications
    slack_test.go:335: Creating ConfigMap...
    slack_test.go:348: Expecting bot message in first channel...
    slack_test.go:360: Expecting no bot message in second channel...
    slack_test.go:366: Updating ConfigMap...
    slack_test.go:373: Expecting bot message in all channels...
    slack_test.go:385: Stopping notifier...
    slack_test.go:393: Getting notifier status from second channel...
    slack_test.go:400: Getting notifier status from first channel...
    slack_test.go:407: Updating ConfigMap once again...
    slack_test.go:414: Ensuring bot didn't write anything new on first channel...
    slack_test.go:420: Expecting bot message on second channel...
    slack_test.go:431: Starting notifier
    slack_test.go:438: Creating and deleting ignored ConfigMap
    slack_test.go:453: Ensuring bot didn't write anything new...
    slack_test.go:458: Deleting ConfigMap
    slack_test.go:463: Expecting bot message on first channel...
    slack_test.go:475: Ensuring bot didn't write anything new on second channel...
=== RUN   TestSlack/Recommendations
    slack_test.go:492: Creating Pod...
    slack_test.go:510: Expecting bot message...
    slack_test.go:569: Cleaning up created Pod...
=== CONT  TestSlack
    slack_test.go:96: Restoring envs for deployment...
    slack_test.go:82: Archiving channel "test-11118aef-682b-41e8-b800-a7c6fdeba05b"...
    slack_test.go:80: Archiving channel "test-8daaab80-f7b7-4cce-9dce-6e883967ee4e"...
--- PASS: TestSlack (352.56s)
    --- PASS: TestSlack/Ping (1.48s)
    --- PASS: TestSlack/Filters_list (1.55s)
    --- PASS: TestSlack/Commands_list (65.83s)
        --- PASS: TestSlack/Commands_list/With_default_cluster (2.54s)
        --- PASS: TestSlack/Commands_list/With_custom_cluster_name (1.52s)
        --- PASS: TestSlack/Commands_list/With_unknown_cluster_name (61.77s)
    --- PASS: TestSlack/Executor (16.41s)
        --- PASS: TestSlack/Executor/Get_Deployment (2.51s)
        --- PASS: TestSlack/Executor/Get_Configmap (1.50s)
        --- PASS: TestSlack/Executor/Get_forbidden_resource (1.49s)
        --- PASS: TestSlack/Executor/Specify_unknown_command (1.48s)
        --- PASS: TestSlack/Executor/Specify_invalid_command (1.47s)
        --- PASS: TestSlack/Executor/Specify_forbidden_namespace (1.47s)
        --- PASS: TestSlack/Executor/Based_on_other_bindings (6.48s)
            --- PASS: TestSlack/Executor/Based_on_other_bindings/Wait_for_Deployment_(the_2st_binding) (1.82s)
            --- PASS: TestSlack/Executor/Based_on_other_bindings/Exec_(the_3rd_binding_which_is_disabled) (1.57s)
            --- PASS: TestSlack/Executor/Based_on_other_bindings/Get_all_Pods_(the_4th_binding) (1.58s)
            --- PASS: TestSlack/Executor/Based_on_other_bindings/Get_all_Deployments_(the_4th_binding) (1.52s)
    --- PASS: TestSlack/Multi-channel_notifications (257.45s)
    --- PASS: TestSlack/Recommendations (1.28s)
PASS
```

**Comment by ezodude** [comment]
@huseyinbabal Can you please hold off merging this until the https://github.com/kubeshop/botkube/pull/723 is merged?

This will help me get the new E2E tests over the line before adding new functionality that requires e2e updates.

**Comment by huseyinbabal** [comment]
Converted to Draft to  decide on some ideas about socket mode flakines

File: comm_config.yaml.tpl [hunk 2]
```diff
@@ -15,6 +15,8 @@ communications:
             sources:
               - 'k8s-events'
       token: 'SLACK_API_TOKEN'
+      botToken: 'SLACK_BOT_TOKEN'
+      appToken: 'SLACK_APP_TOKEN'
```
Comment (mszostok) [inline comm_config.yaml.tpl:19]: please also reset the values here:

```suggestion
      token: "" # SLACK_API_TOKEN
      botToken: "" # SLACK_BOT_TOKEN
      appToken: "" # SLACK_APP_TOKEN
```

**Review by mszostok — APPROVED** [review]
LGTM 👍

**Pull request merged**